From 160184d8cae3c189b16b5c1d03e3202817f8f305 Mon Sep 17 00:00:00 2001
From: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Date: Thu, 7 Sep 2023 12:37:38 +0200
Subject: [PATCH 0537/1695] drivers: add stm32mp1 remoteproc driver

This driver is responsible for configuring the registers and memories of
the remote processor.
- It stores information about memories assigned to the remote processor
  based on the device tree.
- It ensures consistency between the registered memory and the addresses
of the firmware segments to be loaded.
- Additionally, it is responsible for starting and stopping the remote
processor core.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 2f63eaab330afed84703675f9850d4c26fd5cf15 drivers: add stm32mp1 remoteproc driver]
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>
Acked-by: Jerome Forissier <jerome.forissier@linaro.org>
Change-Id: I33a9bc9bc200c575a30e04e49d37d349945a69e5
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/348391
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
---
 core/arch/arm/plat-stm32mp1/conf.mk        |   3 +
 core/drivers/remoteproc/stm32_remoteproc.c | 388 +++++++++++++++++++++
 core/drivers/remoteproc/sub.mk             |   1 +
 core/drivers/sub.mk                        |   1 +
 core/include/drivers/stm32_remoteproc.h    |  70 ++++
 mk/config.mk                               |   4 +
 6 files changed, 467 insertions(+)
 create mode 100644 core/drivers/remoteproc/stm32_remoteproc.c
 create mode 100644 core/drivers/remoteproc/sub.mk
 create mode 100644 core/include/drivers/stm32_remoteproc.h

diff --git a/core/arch/arm/plat-stm32mp1/conf.mk b/core/arch/arm/plat-stm32mp1/conf.mk
index 575f3d407..31fbb34b6 100644
--- a/core/arch/arm/plat-stm32mp1/conf.mk
+++ b/core/arch/arm/plat-stm32mp1/conf.mk
@@ -214,6 +214,9 @@ CFG_MMAP_REGIONS ?= 30
 CFG_DTB_MAX_SIZE ?= (256 * 1024)
 CFG_CORE_ASLR ?= n
 
+CFG_STM32MP_REMOTEPROC ?= n
+CFG_DRIVERS_REMOTEPROC ?= $(CFG_STM32MP_REMOTEPROC)
+
 ifneq ($(CFG_WITH_LPAE),y)
 # Without LPAE, default TEE virtual address range is 1MB, we need at least 2MB.
 CFG_TEE_RAM_VA_SIZE ?= 0x00200000
diff --git a/core/drivers/remoteproc/stm32_remoteproc.c b/core/drivers/remoteproc/stm32_remoteproc.c
new file mode 100644
index 000000000..e4cd6c8b1
--- /dev/null
+++ b/core/drivers/remoteproc/stm32_remoteproc.c
@@ -0,0 +1,388 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2023, STMicroelectronics
+ */
+
+#include <assert.h>
+#include <config.h>
+#include <drivers/rstctrl.h>
+#include <drivers/stm32_remoteproc.h>
+#include <kernel/cache_helpers.h>
+#include <kernel/dt_driver.h>
+#include <kernel/tee_misc.h>
+#include <libfdt.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+
+#define TIMEOUT_US_1MS	U(1000)
+
+/**
+ * struct stm32_rproc_mem - Memory regions used by the remote processor
+ *
+ * @addr:	physical base address from the CPU space perspective
+ * @da:		device address corresponding to the physical base address
+ *		from remote processor space perspective
+ * @size:	size of the region
+ */
+struct stm32_rproc_mem {
+	paddr_t addr;
+	paddr_t da;
+	size_t size;
+};
+
+/**
+ * struct stm32_rproc_instance - rproc instance context
+ *
+ * @cdata:	pointer to the device compatible data
+ * @link:	the node in the rproc_list
+ * @n_regions:	number of memory regions
+ * @regions:	memory regions used
+ * @mcu_rst:	remote processor reset control
+ * @hold_boot:	remote processor hold boot control
+ */
+struct stm32_rproc_instance {
+	const struct stm32_rproc_compat_data *cdata;
+	SLIST_ENTRY(stm32_rproc_instance) link;
+	size_t n_regions;
+	struct stm32_rproc_mem *regions;
+	struct rstctrl *mcu_rst;
+	struct rstctrl *hold_boot;
+};
+
+/**
+ * struct stm32_rproc_compat_data - rproc associated data for compatible list
+ *
+ * @rproc_id:	identifies the remote processor
+ */
+struct stm32_rproc_compat_data {
+	uint32_t rproc_id;
+};
+
+static SLIST_HEAD(, stm32_rproc_instance) rproc_list =
+		SLIST_HEAD_INITIALIZER(rproc_list);
+
+void *stm32_rproc_get(uint32_t rproc_id)
+{
+	struct stm32_rproc_instance *rproc = NULL;
+
+	SLIST_FOREACH(rproc, &rproc_list, link)
+		if (rproc->cdata->rproc_id == rproc_id)
+			break;
+
+	return rproc;
+}
+
+TEE_Result stm32_rproc_start(uint32_t rproc_id)
+{
+	struct stm32_rproc_instance *rproc = stm32_rproc_get(rproc_id);
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (!rproc || !rproc->hold_boot)
+		return TEE_ERROR_GENERIC;
+
+	/*
+	 * The firmware is started by de-asserting the hold boot and
+	 * asserting it back to avoid auto restart on a crash.
+	 * No need to release the MCU reset as it is automatically released by
+	 * the hardware.
+	 */
+	res = rstctrl_deassert_to(rproc->hold_boot, TIMEOUT_US_1MS);
+	if (!res)
+		res = rstctrl_assert_to(rproc->hold_boot, TIMEOUT_US_1MS);
+
+	return res;
+}
+
+static TEE_Result rproc_stop(struct stm32_rproc_instance *rproc)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (!rproc->hold_boot || !rproc->mcu_rst)
+		return TEE_ERROR_GENERIC;
+
+	res = rstctrl_assert_to(rproc->hold_boot, TIMEOUT_US_1MS);
+	if (!res)
+		res = rstctrl_assert_to(rproc->mcu_rst, TIMEOUT_US_1MS);
+
+	return res;
+}
+
+TEE_Result stm32_rproc_stop(uint32_t rproc_id)
+{
+	struct stm32_rproc_instance *rproc = stm32_rproc_get(rproc_id);
+
+	if (!rproc)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return rproc_stop(rproc);
+}
+
+TEE_Result stm32_rproc_da_to_pa(uint32_t rproc_id, paddr_t da, size_t size,
+				paddr_t *pa)
+{
+	struct stm32_rproc_instance *rproc = stm32_rproc_get(rproc_id);
+	struct stm32_rproc_mem *mems = NULL;
+	unsigned int i = 0;
+
+	if (!rproc)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mems = rproc->regions;
+
+	for (i = 0; i < rproc->n_regions; i++) {
+		if (core_is_buffer_inside(da, size, mems[i].da, mems[i].size)) {
+			/*
+			 * A match between the requested DA memory area and the
+			 * registered regions has been found.
+			 * The PA is the reserved-memory PA address plus the
+			 * delta between the requested DA and the
+			 * reserved-memory DA address.
+			 */
+			*pa = mems[i].addr + da - mems[i].da;
+			return TEE_SUCCESS;
+		}
+	}
+
+	return TEE_ERROR_ACCESS_DENIED;
+}
+
+TEE_Result stm32_rproc_map(uint32_t rproc_id, paddr_t pa, size_t size,
+			   void **va)
+{
+	struct stm32_rproc_instance *rproc = stm32_rproc_get(rproc_id);
+	struct stm32_rproc_mem *mems = NULL;
+	unsigned int i = 0;
+
+	if (!rproc)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mems = rproc->regions;
+
+	for (i = 0; i < rproc->n_regions; i++) {
+		if (!core_is_buffer_inside(pa, size, mems[i].addr,
+					   mems[i].size))
+			continue;
+		*va = core_mmu_add_mapping(MEM_AREA_RAM_NSEC, pa, size);
+		if (!*va) {
+			EMSG("Can't map region %#"PRIxPA" size %zu", pa, size);
+			return TEE_ERROR_GENERIC;
+		}
+
+		return TEE_SUCCESS;
+	}
+
+	return TEE_ERROR_ACCESS_DENIED;
+}
+
+TEE_Result stm32_rproc_unmap(uint32_t rproc_id, void *va, size_t size)
+{
+	struct stm32_rproc_instance *rproc = stm32_rproc_get(rproc_id);
+	struct stm32_rproc_mem *mems = NULL;
+	paddr_t pa = virt_to_phys(va);
+	unsigned int i = 0;
+
+	if (!rproc || !pa)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mems = rproc->regions;
+
+	for (i = 0; i < rproc->n_regions; i++) {
+		if (!core_is_buffer_inside(pa, size, mems[i].addr,
+					   mems[i].size))
+			continue;
+
+		/* Flush the cache before unmapping the memory */
+		dcache_clean_range(va, size);
+
+		if (core_mmu_remove_mapping(MEM_AREA_RAM_NSEC, va, size)) {
+			EMSG("Can't unmap region %#"PRIxPA" size %zu",
+			     pa, size);
+			return TEE_ERROR_GENERIC;
+		}
+
+		return TEE_SUCCESS;
+	}
+
+	return TEE_ERROR_ACCESS_DENIED;
+}
+
+static TEE_Result stm32_rproc_get_dma_range(struct stm32_rproc_mem *region,
+					    const void *fdt, int node)
+{
+	const fdt32_t *list = NULL;
+	int ahb_node = 0;
+	int len = 0;
+	int nranges = 0;
+	int i = 0;
+
+	/*
+	 * The match between local and remote processor memory mapping is
+	 * described in the dma-ranges defined by the bus parent node.
+	 */
+	ahb_node = fdt_parent_offset(fdt, node);
+
+	list = fdt_getprop(fdt, ahb_node, "dma-ranges", &len);
+	if (!list) {
+		if (len != -FDT_ERR_NOTFOUND)
+			return TEE_ERROR_GENERIC;
+		/* Same memory mapping */
+		DMSG("No dma-ranges found in DT");
+		region->da = region->addr;
+		return TEE_SUCCESS;
+	}
+
+	if ((len % (sizeof(uint32_t) * 3)))
+		return TEE_ERROR_GENERIC;
+
+	nranges = len / sizeof(uint32_t);
+
+	for (i = 0; i < nranges; i += 3) {
+		uint32_t da = fdt32_to_cpu(list[i]);
+		uint32_t pa = fdt32_to_cpu(list[i + 1]);
+		uint32_t size = fdt32_to_cpu(list[i + 2]);
+
+		if (core_is_buffer_inside(region->addr, region->size,
+					  pa, size)) {
+			region->da = da + (region->addr - pa);
+			return TEE_SUCCESS;
+		}
+	}
+
+	return TEE_ERROR_BAD_PARAMETERS;
+}
+
+/* Get device tree memory regions reserved for the Cortex-M and the IPC */
+static TEE_Result stm32_rproc_parse_mems(struct stm32_rproc_instance *rproc,
+					 const void *fdt, int node)
+{
+	const fdt32_t *list = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct stm32_rproc_mem *regions = NULL;
+	int len = 0;
+	int n_regions = 0;
+	int i = 0;
+
+	list = fdt_getprop(fdt, node, "memory-region", &len);
+	if (!list) {
+		EMSG("No memory regions found in DT");
+		return TEE_ERROR_GENERIC;
+	}
+
+	n_regions = len / sizeof(uint32_t);
+
+	regions = calloc(n_regions, sizeof(*regions));
+	if (!regions)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	for (i = 0; i < n_regions; i++) {
+		int pnode = 0;
+
+		pnode = fdt_node_offset_by_phandle(fdt, fdt32_to_cpu(list[i]));
+		if (pnode < 0) {
+			res = TEE_ERROR_GENERIC;
+			goto err;
+		}
+
+		regions[i].addr = fdt_reg_base_address(fdt, pnode);
+		regions[i].size = fdt_reg_size(fdt, pnode);
+
+		if (regions[i].addr <= 0 || regions[i].size <= 0) {
+			res = TEE_ERROR_GENERIC;
+			goto err;
+		}
+
+		res = stm32_rproc_get_dma_range(&regions[i], fdt, node);
+		if (res)
+			goto err;
+
+		if (!regions[i].addr || !regions[i].size) {
+			res = TEE_ERROR_BAD_PARAMETERS;
+			goto err;
+		}
+
+		DMSG("register region %#"PRIxPA" size %#zx",
+		     regions[i].addr, regions[i].size);
+	}
+
+	rproc->n_regions = n_regions;
+	rproc->regions = regions;
+
+	return TEE_SUCCESS;
+
+err:
+	free(regions);
+
+	return res;
+}
+
+static void stm32_rproc_cleanup(struct stm32_rproc_instance *rproc)
+{
+	free(rproc->regions);
+	free(rproc);
+}
+
+static TEE_Result stm32_rproc_probe(const void *fdt, int node,
+				    const void *comp_data)
+{
+	struct stm32_rproc_instance *rproc = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	rproc = calloc(1, sizeof(*rproc));
+	if (!rproc)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	rproc->cdata = comp_data;
+
+	res = stm32_rproc_parse_mems(rproc, fdt, node);
+	if (res)
+		goto err;
+
+	res = rstctrl_dt_get_by_name(fdt, node, "mcu_rst", &rproc->mcu_rst);
+	if (res)
+		goto err;
+
+	res = rstctrl_dt_get_by_name(fdt, node, "hold_boot", &rproc->hold_boot);
+	if (res)
+		goto err;
+
+	/* Ensure that the MCU is HOLD */
+	if (rproc->mcu_rst) {
+		res = rproc_stop(rproc);
+		if (res)
+			goto err;
+	}
+
+	/*
+	 * The memory management should be enhance with firewall
+	 * mechanism to map the memory in secure area for the firmware
+	 * loading and then to give exclusive access right to the
+	 * coprocessor (except for the shared memory).
+	 */
+	IMSG("Warning: the remoteproc memories are not protected by firewall");
+
+	SLIST_INSERT_HEAD(&rproc_list, rproc, link);
+
+	return TEE_SUCCESS;
+
+err:
+	stm32_rproc_cleanup(rproc);
+	return res;
+}
+
+static const struct stm32_rproc_compat_data stm32_rproc_m4_compat = {
+	.rproc_id = STM32_M4_RPROC_ID,
+};
+
+static const struct dt_device_match stm32_rproc_match_table[] = {
+	{
+		.compatible = "st,stm32mp1-m4-tee",
+		.compat_data = &stm32_rproc_m4_compat,
+	},
+	{ }
+};
+
+DEFINE_DT_DRIVER(stm32_rproc_dt_driver) = {
+	.name = "stm32-rproc",
+	.match_table = stm32_rproc_match_table,
+	.probe = &stm32_rproc_probe,
+};
diff --git a/core/drivers/remoteproc/sub.mk b/core/drivers/remoteproc/sub.mk
new file mode 100644
index 000000000..4eb5ca3f2
--- /dev/null
+++ b/core/drivers/remoteproc/sub.mk
@@ -0,0 +1 @@
+srcs-$(CFG_STM32MP_REMOTEPROC) += stm32_remoteproc.c
diff --git a/core/drivers/sub.mk b/core/drivers/sub.mk
index c8fce5e82..a6ef8e93a 100644
--- a/core/drivers/sub.mk
+++ b/core/drivers/sub.mk
@@ -96,6 +96,7 @@ subdirs-$(CFG_DRIVERS_NVMEM) += nvmem
 subdirs-$(CFG_DRIVERS_PINCTRL) += pinctrl
 subdirs-$(CFG_DRIVERS_REGULATOR) += regulator
 subdirs-$(CFG_DRIVERS_RSTCTRL) += rstctrl
+subdirs-$(CFG_DRIVERS_REMOTEPROC) += remoteproc
 subdirs-$(CFG_SCMI_MSG_DRIVERS) += scmi-msg
 subdirs-y += imx
 subdirs-y += pm
diff --git a/core/include/drivers/stm32_remoteproc.h b/core/include/drivers/stm32_remoteproc.h
new file mode 100644
index 000000000..8cb1efd86
--- /dev/null
+++ b/core/include/drivers/stm32_remoteproc.h
@@ -0,0 +1,70 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2023, STMicroelectronics - All Rights Reserved
+ */
+
+#ifndef __DRIVERS_STM32_REMOTEPROC_H
+#define __DRIVERS_STM32_REMOTEPROC_H
+
+#include <stdint.h>
+#include <tee_api_types.h>
+#include <types_ext.h>
+
+/* IDs of the supported remote processors*/
+#define STM32_M4_RPROC_ID 0
+
+/*
+ * stm32_rproc_get() - get the rproc handle associated to a remote processor ID
+ * @rproc_id unique identifier of the remote processor
+ * Return a pointer to the rproc firmware handle related to @rproc_id or NULL.
+ */
+void *stm32_rproc_get(uint32_t rproc_id);
+
+/*
+ * stm32_rproc_da_to_pa() - Convert the coprocessor device address to a CPU
+ *                          physical address.
+ * @rproc_id	unique identifier of the remote processor
+ * @da		device memory address from the remote processor space
+ *		perspective.
+ * @size	size of the memory
+ * @pa		Output CPU physical address associated to @da.
+ * Return TEE_SUCCESS or appropriate error.
+ */
+TEE_Result stm32_rproc_da_to_pa(uint32_t rproc_id, paddr_t da, size_t size,
+				paddr_t *pa);
+
+/*
+ * stm32_rproc_map() - map the physical address if valid
+ * @rproc_id	unique identifier of the remote processor
+ * @pa		physical address from the CPU space perspective
+ * @size	size of the memory
+ * @va		Output CPU virtual address associated to @pa.
+ * Return TEE_SUCCESS or appropriate error.
+ */
+TEE_Result stm32_rproc_map(uint32_t rproc_id, paddr_t pa, size_t size,
+			   void **va);
+
+/*
+ * stm32_rproc_unmap() - ummap the virtual address mapped with stm32_rproc_map
+ * @rproc_id	unique identifier of the remote processor
+ * @va		virtual address
+ * @size	size of the memory
+ * Return TEE_SUCCESS or appropriate error.
+ */
+TEE_Result stm32_rproc_unmap(uint32_t rproc_id, void *va, size_t size);
+
+/*
+ * stm32_rproc_start() - start the remote processor core
+ * @rproc_id	unique identifier of the remote processor
+ * Return TEE_SUCCESS or appropriate error.
+ */
+TEE_Result stm32_rproc_start(uint32_t rproc_id);
+
+/*
+ * stm32_rproc_stop() - stop the remote processor core
+ * @rproc_id	unique identifier of the remote processor
+ * Return TEE_SUCCESS or appropriate error.
+ */
+TEE_Result stm32_rproc_stop(uint32_t rproc_id);
+
+#endif /* __DRIVERS_STM32_REMOTEPROC_H */
diff --git a/mk/config.mk b/mk/config.mk
index ffd7d1a5d..418e12881 100644
--- a/mk/config.mk
+++ b/mk/config.mk
@@ -1072,6 +1072,10 @@ ifeq (y-y,$(CFG_CORE_PREALLOC_EL0_TBLS)-$(CFG_WITH_PAGER))
 $(error "CFG_WITH_PAGER can't support CFG_CORE_PREALLOC_EL0_TBLS")
 endif
 
+
+# CFG_DRIVERS_REMOTEPROC, when enabled, embeds support for remote processor
+# management including generic DT bindings for the configuration.
+CFG_DRIVERS_REMOTEPROC ?= n
 # User TA runtime context dump.
 # When this option is enabled, OP-TEE provides a debug method for
 # developer to dump user TA's runtime context, including TA's heap stats.
-- 
2.47.2

